### ISSUE TITLE
Add QuadTree/Octree or insertions to KDTree

### ISSUE THREAD

**Issue by bthayer2365** [scipy/scipy#10551]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to insert to a nearest neighbor data structure.

**Describe the solution you'd like**
For the project I'm working on, I frequently need to find the nearest data points within a certain range. The points that I'm working with are 2d, so it seems like a quadtree would be a beneficial data structure to have for this task. If we have a quadtree, it would make sense to have an octree too.

**Describe alternatives you've considered**
Right now, it is fastest to just use numpy and calculate the distances between all points and the point that I'm interested in, but it feels like there should be a better way to do this. 

Ex:
```
points = np.random.random((10000, 2))
point= np.random.random(2)
threshold = 0.02
distances = np.linalg.norm(points - point, axis=1)
neighbors = distances > threshold
```
KDTrees don't allow insertion/deletion, and I'd rather not have a dependency for a separate quadtree library which runs the risk of conflicting dependencies and not being maintained. I could implement one myself, but if I'm going through that effort, it feels like it should be on scipy.

**Additional context (e.g. screenshots)**

**Comment by rgommers** [comment]
A good quadtree and/or octree implementation would be a great addition to `scipy.spatial` imho. There may be a library out there already though. Have you searched for existing implementations in Python/C/C++?

**Comment by bthayer2365** [comment]
I found this question on stack overflow.

https://stackoverflow.com/questions/2298517/are-any-of-these-quad-tree-libraries-any-good

It seems like the best implementation is in pygame http://www.pygame.org/wiki/QuadTree. I haven't used pygame, but it seems like this isn't even an importable library and they're suggesting that you copy the code yourself.

It seems like there are many QuadTree implementations in C++/C. Google finds a bunch of examples
https://gist.github.com/shadmar/f0562eb9e2abec5637b3
https://stackoverflow.com/questions/4544928/quadtree-explanation-and-c-implementation
It seems like there's many more, but I haven't tested any of them

**Comment by bthayer2365** [comment]
If I were to work on this, should I just copy them and include a reference to the source code I'm basing it off of or should we import the existing libraries? I can always go based off of Wikipedia and other sources and just write it from scratch. I think it probably makes sense to have both a Python implementation and a C/C++ implementation given that that's the case for the KDTrees.

**Comment by tylerjereddy** [comment]
>  I think it probably makes sense to have both a Python implementation and a C/C++ implementation given that that's the case for the KDTrees.

I think just the faster implementation with an appropriate Python API above it might be preferred, though I may be wrong. I feel like almost everyone just uses `cKDTree` now instead of the Python version, though perhaps there are interesting exceptions.

**Comment by sturlamolden** [comment]
The Python version of kd-tree will be removed and just kept as a fancy test class. It is just a historical relic.

**Comment by sturlamolden** [comment]
I would like to advice though, that the key to performance when querying a spatial data structure is in the details.  It is very easy to write a kd-tree, ball-tree, quad-tree, oct-tree, etc. that will be slow as hell, but otherwise correct. If you look at cKDTree you will see that the query code keeps nodes and neighbors in heaps, prefetches data points, uses a specialized memory allocator, etc. When building the tree, the cut-offs are decided intelligently, using quickselect or sliding midpoints. This is the kind of machinery you need to get a fast query. If you do not, chances are your quad-tree will be slower than just running cKDTree on the same data set.

**Comment by bthayer2365** [comment]
Yeah, I took a look at the code and was caught off guard by KDTree's
complexity. Now since you seem to understand this better than I do, I want
to ask: how much of the machinery in the KDTree can be reused? And how much
faster should we expect a QuadTree to be? In terms of implementation, I was
already thinking of possibly not even using a QuadTree, instead using a
tree-pyramid with a different grid structure, having leaf nodes with many
points in order to take advantage of the branch predictor, etc. It might
not even make sense to use a tree-pyramid at all, but at least for the
special cases of 2d and 3d, it seems like there should be more
optimizations we can make to produce a faster data structure than a KDTree,
which was built to handle the more general problem.

**Comment by sturlamolden** [comment]
Well, one can probably reuse a lot. A kd-tree is fundamentally related to a quad-tree (and an oct-tree). 

With a 2D space it goes like this:

- At each node in a kd-tree you split the space into two rectangles along one dimension. Each node will then have two children. Then you alternate which dimension to split, according to some heuristic.

- At each node in a quad-tree you split the space into four rectangles, by splitting along both dimensions. Each node will then have four children.

The machinery will thus be roughly the same, e.g. using heaps as priority queues for deciding which node to trace first. A quad tree will theoretically be faster, but requires a bit more logic.
